WireDomain/Sources/WireDomain/Repositories/Message/MessageLocalStore.swift

```diff
@@ -17,18 +17,154 @@
 //
 
 import CoreData
+import WireCryptobox
 import WireDataModel
 
 // sourcery: AutoMockable
 /// Facilitate access to message related domain objects.
 public protocol MessageLocalStoreProtocol {
 
-    func addSystemMessageToConversation(
-        messageType: MessageType,
+    /// Adds a system message to a given conversation.
+    /// - Parameters:
+    ///     - messageType: The type of system message to add.
+    ///     - conversationID: The conversation id.
+    ///     - conversationDomain: The conversation domain.
+
+    func addSystemMessage(
+        messageType: SystemMessageType,
         conversationID: UUID,
         conversationDomain: String?
     ) async
 
+    /// Adds a message text to a given conversation.
+    /// - Parameters:
+    ///     - message: The message to add.
+    ///     - conversation: The conversation to add the message text to.
+    ///     - senderID: The message sender id.
+    ///     - senderDomain: The message sender domain.
+    ///     - senderClientID: The sender client id.
+    ///     - date: The date the message was created.
+    ///     - logAttributes: Attributes to add to the log.
+
+    func addTextMessage(
+        _ message: GenericMessage,
+        in conversation: ZMConversation,
+        senderID: UUID,
+        senderDomain: String,
+        senderClientID: String?,
+        date: Date,
+        logAttributes: LogAttributes
+    ) async
+
+    /// Checks whether a message can be added to the conversation.
+    /// - Parameters:
+    ///     - conversation: The conversation to add the message to.
+    ///     - senderID: The message sender id.
+    ///     - logAttributes: Attributes to add to the log.
+
+    func canAddMessage(
+        conversation: ZMConversation,
+        senderID: UUID,
+        logAttributes: LogAttributes
+    ) async -> Bool
```

Comment:
Does this necessarily have to be part of the protocol? Could it not be an internal check that, when fails, results in an error beIng thrown?

Reply:
I think so yes because we don't directly call the message local store right after, we're doing some work before that, and I think we don't want to this extra work if in the end we cannot add the message. that's why we make this check early on.